### DISCUSSION TITLE
Why are composited images different on different Operating Systems?

### DISCUSSION THREAD

**Discussion by lukestoward** [libvips/libvips#2782] [Q&A]
This might be an obvious question, but I don't have a lot of experience with image transformation so I'd appreciate any help in understanding this.

When I run the following commands on OSX and Amazon Linux respectively:

```bash
# run on OSX
vips composite "base_image.png overlay.png" vips_mac.png 5

# run on Amazon Linux 2 Docker image
vips composite "base_image.png overlay.png" vips_linux.png 5
```

And compare the output images via:

```bash
compare -compose src vips_linux.png vips_mac.png diff.png
```

I see multiple pixels highlighted in red signalling a difference. The different pixels are located on the overlay graphic.  **Why are the output images different?** The version of `vips` is `vips-8.12.2-Tue Jan 25 09:34:32 UTC 2022` in both cases.

The `overlay.png` is a logo with an alpha channel and the `base_image` is a larger image with no alpha channel.

**Comment by jcupitt** [comment]
Hi @lukestoward,

At a guess, they are using different code paths. `composite` has a vector mode when built with gcc, but will fall back to a plain C path when built with clang. Hopefully the differences are small!

You could measure the peak difference with:

```
vips subtract vips_mac.png vips_linux.png t1.v
vips abs t1.v t2.v
vips max t2.v
```

**Reply by lukestoward** [reply]
Thanks for the information @jcupitt.

Running those commands returns `1.000000`.

Do you know if there is a way I can force the two vips instances to behave the same way?

**Reply by jcupitt** [reply]
That's a tiny difference, just rounding, I would be tempted to ignore it.

The vector path is just for uchar, ushort and float images, so casting your PNGs to something like uint for composite would force vector off. Something like:

```
vips cast base_image.png t1.v uint
vips cast overlay.png t2.v uint
vips composite2 t1.v t2.v vips_linux.png over
```

Should work.
